### ISSUE TITLE
Unable to Reproduce Paper Results After Dozens of Experiments - Code Organization Issues and Missing Training Details

### ISSUE THREAD

**Issue by bi32** [ShanghaiTech-IMPACT/3D-MedDiffusion#4]
Hi, thank you for sharing this excellent work on 3D medical image generation.

I have been attempting to reproduce the results from your paper over the past few weeks using 8× H100 GPUs, conducting **extensive experiments (over 30 different configuration combinations)**, but I consistently fail to achieve the generation quality reported in the paper. I have tried various hyperparameter combinations (learning rates, batch sizes, discriminator start timing, loss weights, etc.), but the results remain far below what your paper demonstrates. I suspect the publicly released code may differ from the version actually used in the paper. Below are the specific issues I have identified:

### 1. Unable to Reproduce Reported Results

Despite having substantial computational resources (8×H100 GPUs) and conducting systematic experiments:

**Configuration combinations I have tried include:**
- Learning rates: 1e-4, 3e-4, 5e-4, 1e-3 (paper default: 3e-4)
- Batch sizes: 8, 16, 24, 32, 48 (single-GPU and multi-GPU)
- Discriminator start iterations: 5000, 10000, 12000, 15000, 20000 steps
- Loss weight combinations:
  - `l1_weight`: 2.0, 4.0, 8.0
  - `perceptual_weight`: 1.0, 2.0, 4.0
  - `volume_gan_weight`: 1.0, 2.0, 4.0
  - `gan_feat_weight`: 2.0, 4.0, 8.0
- Training steps: 50K, 100K, 200K, 500K, 1M steps
- Different perceptual loss configurations (3D vs 2D)

**Yet the generation quality consistently remains far below the paper's demonstrations.**

Could you please provide:
- **Exact dataset details** used for training (specific modality distribution, anatomical regions, preprocessing pipeline, number of samples)?
- Whether there are **critical tricks not implemented in the code** (special data augmentation, normalization strategies, warmup strategies, etc.)?
- **Complete configuration files for paper experiments** and expected intermediate metrics (e.g., what should recon_loss be at 10K steps? What range should perplexity be in?)

### 2. Code Version Confusion - Multiple Implementations Coexist

The repository contains **numerous different versions of core files**, and these files are all committed in `__pycache__`, indicating they were all used at some point:

**4 versions of PatchVolume:**
```
AutoEncoder/model/__pycache__/PatchVolume.cpython-311.pyc
AutoEncoder/model/__pycache__/PatchVolume4x.cpython-311.pyc
AutoEncoder/model/__pycache__/PatchVolume_new.cpython-311.pyc
AutoEncoder/model/__pycache__/PatchVolume_old.cpython-311.pyc
```

**2 versions of BiFlowNet:**
```
ddpm/__pycache__/BiFlowNet.cpython-311.pyc
ddpm/__pycache__/BiFlowNetv5.cpython-311.pyc
```

**Multiple versions of VQGAN dataset:**
```
dataset/vqgan.py
dataset/vqgan_4x.py
(plus vqgan_new.pyc, vqgan_memoryinfo.pyc in __pycache__)
```

**Questions:**
- What are the differences between these versions?
- Which version was actually used in the paper experiments?
- Why keep so many versions? Are they leftover code from different experimental stages?
- Which version is being called by current `import` statements?

### 3. Missing Early Stopping Mechanism

I examined all training code and found:

**Stage 1 (VQGAN) training:**
```python
# train/train_PatchVolume.py
callbacks.append(ModelCheckpoint(monitor='val/recon_loss', save_top_k=3, mode='min'))
```
- Although `val/recon_loss` is monitored, there is **no early stopping callback**
- Only saves top-3 checkpoints, but training runs until `max_steps=1000000`
- Config file sets 1 million steps with no early termination mechanism

**Stage 3 (BiFlowNet) training:**
```python
# train/train_BiFlowNet_SingleRes.py:327
# scheduler.step()  # This line is commented out!
```
- **Learning rate scheduler is commented out**, learning rate remains fixed
- Also **no early stopping**
- Training runs for full `args.epochs=1000` epochs

**Questions:**
- Did the paper's actual training really run for the full 1 million steps? Or were there unpublished early stopping criteria?
- Why is the learning rate scheduler commented out? Was learning rate decay used in the paper?
- Are there criteria for stopping when validation performance stops improving?

### 4. Inconsistent Configuration Parameters and Missing Documentation

**Resolution normalization factor confusion:**
```python
# dataset/Singleres_dataset.py:49
return latent, torch.tensor(int(cls_idx)), torch.tensor(self.resolution)/64.0
```
- Why divide by 64.0? Where does this number come from?
- For 4x downsampling (192→48), dividing by 64 gives 0.75 - what's the physical meaning?
- For 8x downsampling (256→32), dividing by 64 gives 0.5 - why use the same normalization factor for different compression ratios?

**Inconsistent discriminator start timing:**
```yaml
PatchVolume_4x.yaml: discriminator_iter_start: 12000
PatchVolume_8x.yaml: discriminator_iter_start: 20000
PatchVolume_4x_s2.yaml: discriminator_iter_start: 0
PatchVolume_8x_s2.yaml: discriminator_iter_start: 0
```
- Why does Stage 2 start discriminator from 0?
- What's the rationale for different GAN training strategies between stages?

**Unclear input sizes:**
- Paper abstract says "up to 512×512×512", but code seems to use patch-based training
- What's the actual input size range during training?
- Is there a fixed input size, or multi-resolution training?

### 5. Git History Shows Code Was Uploaded All at Once

Examining the repository's commit history:
```
31930fa 2025-08-05 1  (Added 4643 lines of code and all __pycache__ files at once)
51e09e6 2025-05-15 Update README.md
ec2103b 2025-05-15 pages uploading
...
2a63cd8 2024-12-16 Create README.md (initial commit)
```

**Issues:**
- Latest commit added all code and compiled cache files at once
- Commit messages are meaningless ("1", "Update README.md", etc.)
- No clear development history to trace which versions correspond to the paper
- This suggests code was directly copied from an internal project, possibly containing mixed experimental versions

### 6. Missing Critical Training Details

**Data preprocessing:**
- What's the intensity normalization strategy for medical images?
- Is special windowing/clipping used?
- Is preprocessing different for different modalities (CT vs MRI)?

**Training stability:**
- Is gradient clipping used?
- Are there special warm-up strategies?
- What's the synchronization strategy for multi-GPU training (I see code uses `ddp_find_unused_parameters_true`)?

**Sampling and evaluation:**
- How many DDPM/DDIM sampling steps are used during inference?
- How are quantitative metrics (FID, SSIM, etc.) in the paper calculated?
- Where is the evaluation code?

### Requests

Given the issues above, could you please provide:

1. **The exact code version used for paper experiments** (ideally as a separate branch or tag)
2. **Complete configuration files** with exact values for all hyperparameters
3. **Expected training curves and intermediate metrics** (e.g., loss values at 10K steps, perplexity range at 100K steps)
4. **Clarification on which code files are the paper version** and which are experimental leftovers
5. **Actual training duration** (was it really trained for 1 million steps?)
6. **Minimal reproducible example** with expected outputs

I am very eager to reproduce your excellent work, but the current state of the code makes it difficult for me to determine where the problem lies - whether it's my experimental setup that's problematic, or whether the publicly released code genuinely differs from the paper experiments.

Thank you for your time and assistance! Looking forward to your response.

**Comment by bi32** [comment]
Minor formatting comments can be skipped; questions on training parameters are blocking and need confirmation. Thanks!

**Comment by bi32** [comment]
Currently, my approach is to use cubic interpolation to build an adapter that exactly matches your input so I don’t change any model dimensions. I also kept only a single class to avoid altering any model structure, and then retrained. I’m using the ADNI dataset; I cleaned it by removing all non‑brain content, performed registration, and applied normalization. 

I’m encountering two generation issues: visible seams and some blurriness (these persist despite extensive tuning). Do you have a recommended early‑stopping criterion for your VQ‑VAE? Around 50 epochs the perceptual loss becomes very low, so I saved a checkpoint there; by 100 epochs the pixel‑wise loss is very low, and I saved that as well, but both checkpoints still produce unsatisfactory results.

I am currently trying to train for one million steps to see if I should continue training, but if I continue training, it seems that the perplexity will drop. I would record the results to see if they are similar to yours, and then find a good save point.

OK, I will try your suggestion. Thanks

**Comment by Mors20** [comment]
## 6

Data preprocessing:
You can refer to the code I posted in Question 1. The normalization process is the same across different modalities, normalizing to (0,1) , but the cropping and resampling differ because each modality has different shapes and spacing ranges.

Training stability:
The training uses PyTorch DistributedDataParallel (DDP). Setting find_unused_parameters=True allows DDP to handle models where some parameters are not used in every forward pass, which occurs due to frozen parameters during Stage 2 training of VQ-GAN.

Sampling and evaluation 
For DDPM, we use 1,000 steps; please refer to my inference code.
For FID calculation, I generated 1,000 images for each competing method as well as for our method, and then computed each method’s FID using the real images as the ground truth distribution.
For SSIM, it is used for downstream tasks and calculated between the reconstructed images and the ground truth.
I compute the MS-SSIM pairwise for the generated images.

Evaluation code
I do not maintain the evaluation code; for reference,  please see https://github.com/batmanlab/HA-GAN

**Comment by bi32** [comment]
<img width="1414" height="1125" alt="Image" src="https://github.com/user-attachments/assets/e5f66911-b7ad-4061-af9a-18da5ec46f29" />

How big is this dataset? It is currently not available for public download. I would like to know how many steps equal to one epoch. Thank you.

**Comment by Mors20** [comment]
> <img alt="Image" width="1414" height="1125" src="https://private-user-images.githubusercontent.com/207951121/496976320-e5f66911-b7ad-4061-af9a-18da5ec46f29.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk0NzE4NzcsIm5iZiI6MTc1OTQ3MTU3NywicGF0aCI6Ii8yMDc5NTExMjEvNDk2OTc2MzIwLWU1ZjY2OTExLWI3YWQtNDA2MS1hZjlhLTE4ZGE1ZWM0NmYyOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAwM1QwNjA2MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOWYwN2IzNmUxMGZiN2Q0ODdmNjdkMzQxYjdkZTIxOTA0ZmU2YjIxZDE1YjdiY2U2ZTgwZDUzNDExOWQyZDM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bukHVsp-kyGRen2LeNqitH0zP-PbRqZX2dZa0_T-1yA">
> How big is this dataset? It is currently not available for public download. I would like to know how many steps equal to one epoch. Thank you.

There are approximately 12,000 images. I train with a batch size of 16, which results in about 750 steps per epoch.

**Comment by bi32** [comment]
<img width="339" height="729" alt="Image" src="https://github.com/user-attachments/assets/4bd67e59-86b0-4c4f-b499-d62135b7cde7" />

I have a question about the Stage 2 fine-tuning step (train_PatchVolume_stage2.py). After completing Stage 1 patch-based training, is Stage 2 (decoder fine-tuning on full images) necessary? What improvements does it provide in terms of reconstruction quality? Can I skip it and go directly to latent generation and diffusion training, or would that significantly impact final generation quality? Thanks!

**Comment by Etine132** [comment]
@Mors20 Hello, I saw that you mentioned that

> (2) For a simple reproduction pipeline, you can:
> Resample all scans in LIDC-IDRI to spacing (1.25, 1.25, 1.25)
> Clip intensities to (-1000 HU, 99.9% quantile)
> Normalize values to (0, 1)

but in the project readme it said that all training images should be normalized to [-1,1], so should  i normalize them to (-1,1) or (0,1)?
